src/casts/specify-representation.rkt

```diff
@@ -443,33 +412,75 @@ but a static single assignment is implicitly maintained.
     (define mk-tuple-crcn-label (Code-Label mk-tuple-crcn))
     (define mk-crcn-label (Code-Label mk-crcn))
     (define mk-tuple-crcn-c : D0-Code
-      (code$ (t1 t2 l i count)
-        (If (op$ = i count)
-            (begin$
-              (assign$ crcn
-                (op$ Alloc (sr-plus count COERCION-TUPLE-ELEMENTS-OFFSET)))
-              (assign$ tagged-count
-                (op$ + (op$ %<< count COERCION-SECOND-TAG-SHIFT)
-                     COERCION-TUPLE-SECOND-TAG))
-              (sr-array-set! crcn COERCION-TUPLE-COUNT-INDEX tagged-count)
-              (sr-tag-value crcn COERCION-MEDIATING-TAG))
-            (begin$
-              (assign$ t1a
-                (sr-tagged-array-ref
-                 t1 TYPE-TUPLE-TAG (sr-plus TYPE-TUPLE-ELEMENTS-OFFSET i)))
-              (assign$ t2a
-                (sr-tagged-array-ref
-                 t2 TYPE-TUPLE-TAG (sr-plus TYPE-TUPLE-ELEMENTS-OFFSET i)))
-              (assign$ val-crcn (App-Code mk-crcn-label `(,t1a ,t2a ,l)))
-              (assign$ tmp-crcn
-                (App-Code
-                 mk-tuple-crcn-label `(,t1 ,t2 ,l ,(sr-plus (Quote 1) i) ,count)))
-              (sr-tagged-array-set!
-               tmp-crcn
-               COERCION-MEDIATING-TAG
-               (sr-plus COERCION-TUPLE-ELEMENTS-OFFSET i)
-               val-crcn)
-              tmp-crcn))))
+      (code$ (t1 t2 l)
+        (assign$ t1-count
+          (sr-tagged-array-ref t1 TYPE-TUPLE-TAG TYPE-TUPLE-COUNT-INDEX))
+        (assign$ t2-count
+          (sr-tagged-array-ref t2 TYPE-TUPLE-TAG TYPE-TUPLE-COUNT-INDEX))
+        (cond$
+         [(op$ = t1-count t2-count)
+          ;; if the two tuple types have the same size, they are not equal
```

Comment:
How do we know they are not equal? If I understand correctly it is because we know the contexts in which this function if called. If that is the case then we should document that we rely on this invariant.

Reply:
they are not equal because of type hashconsing.

Reply:
I got rid of the branching.